src/fsharp/FxResolver.fs

```diff
@@ -1,50 +1,180 @@
+type internal FxResolver(useDotNetFramework: bool option, projectDir: string, m: range) =
+
+    static let isWindows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows)
+
+    static let dotnet =
+        if isWindows then "dotnet.exe" else "dotnet"
+
+    static let dotnetHostPath =
+        // How to find dotnet.exe --- woe is me; probing rules make me sad.
+        // Algorithm:
+        // 1. Look for DOTNET_HOST_PATH environment variable
+        //    this is the main user programable override .. provided by user to find a specific dotnet.exe
+        // 2. Probe for are we part of an .NetSDK install
+        //    In an sdk install we are always installed in:   sdk\3.0.100-rc2-014234\FSharp
+        //    dotnet or dotnet.exe will be found in the directory that contains the sdk directory
+        // 3. We are loaded in-process to some other application ... Eg. try .net
+        //    See if the host is dotnet.exe ... from netcoreapp3.1 on this is fairly unlikely
+        // 4. If it's none of the above we are going to have to rely on the path containing the way to find dotnet.exe
+        //
+        match (Environment.GetEnvironmentVariable("DOTNET_HOST_PATH")) with
+        | value when not (String.IsNullOrEmpty(value)) ->
+            value                           // Value set externally
+        | _ ->
+            // Probe for netsdk install, dotnet. and dotnet.exe is a constant offset from the location of System.Int32
+            if isRunningOnCoreClr then
+                let dotnetLocation =
+                    let dotnetApp =
+                        let platform = Environment.OSVersion.Platform
+                        if platform = PlatformID.Unix then "dotnet" else "dotnet.exe"
+                    let assemblyLocation = Path.GetDirectoryName(typeof<Int32>.GetTypeInfo().Assembly.Location)
+                    Path.GetFullPath(Path.Combine(assemblyLocation, "../../..", dotnetApp))
+
+                if File.Exists(dotnetLocation) then
+                    dotnetLocation
+                else
+                    let main = Process.GetCurrentProcess().MainModule
+                    if main.ModuleName ="dotnet" then
+                        main.FileName
+                    else
+                        dotnet
+            elif isWindows then
+                let loc = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles),"dotnet","dotnet.exe")
+                if File.Exists(loc) then
+                    loc
+                else dotnet
+            else
+                dotnet
+
+    static let desiredDotNetSdkVersionForDirectoryCache = ConcurrentDictionary<string, string>()
 
-    let inline ifEmptyUse alternative filename = if String.IsNullOrWhiteSpace filename then alternative else filename
+    /// Find the relevant sdk version by running `dotnet --version` in the script/project location,
+    /// taking into account any global.json
+    let tryGetDesiredDotNetSdkVersionForDirectory() =
```

Comment:
I had imagined that this would not be done at this level.  It seemed to me that this should be done at the point that the editor, creates a dummy project file with default references and command line options.  It feels like this is way to low in the stack to be getting this information for the editor, which is why fsi --info seemed like a good choice for getting those values.